src/main/java/com/enderio/base/common/integrations/jei/extension/ShapedEntityStorageCategoryExtension.java

```diff
@@ -24,38 +29,99 @@ public ShapedEntityStorageCategoryExtension() {}
     @Override
     public void setRecipe(RecipeHolder<ShapedEntityStorageRecipe> recipeHolder, IRecipeLayoutBuilder builder, ICraftingGridHelper craftingGridHelper, IFocusGroup focuses) {
 
-        var allCapturableEntities = EntityCaptureUtils.getCapturableEntities();
-
+        Optional<IFocus<ItemStack>> output = focuses.getItemStackFocuses(OUTPUT).findFirst();
+        Optional<IFocus<ItemStack>> input = focuses.getItemStackFocuses(INPUT).filter(f -> f.getTypedValue().getItemStack().get().is(EIOTags.Items.ENTITY_STORAGE)).findFirst();
         ShapedEntityStorageRecipe recipe = recipeHolder.value();
         ItemStack resultItem = EnderIOJEI.getResultItem(recipe);
-        List<ItemStack> results = allCapturableEntities.stream().map(e -> {
-            ItemStack result = resultItem.copy();
+        List<List<ItemStack>> inputs = recipe.getIngredients().stream()
+            .map(ingredient ->
+                Arrays.stream(ingredient.getItems()).<ItemStack>mapMulti(
+                    (ingredientItem, consumer) -> consumer.accept(ingredientItem))
+                    .toList())
+            .toList();
+        List<ItemStack> results = List.of(resultItem);
+        boolean noData = true;
 
-            if (result.is(EIOTags.Items.ENTITY_STORAGE)) {
-                result.set(EIODataComponents.STORED_ENTITY, StoredEntityData.of(e));
+        if (input.isPresent()) {
```

Comment:
I'm struggling to follow the new logic here - could you maybe explain what you've changed and why?

Reply:
This logic is to make it so the input soul and output soul match when searching. Otherwise they keep cycling and are out of sync.